src/search/tasks/negated_axioms_task.cc

```diff
@@ -0,0 +1,323 @@
+    for (const FactProxy &goal: task_proxy.get_goals()) {
+        if (goal.get_variable().is_derived() && goal.get_value() == 1) { // TODO: fix the hardcoded 1
+            needed_negatively.insert(goal.get_pair().var);
+        }
+    }
+    for (OperatorProxy op: task_proxy.get_operators()) {
+        for (FactProxy condition: op.get_preconditions()) {
+            if (condition.get_variable().is_derived() && condition.get_value() == 1) {  // TODO: fix the hardcoded 1
+                needed_negatively.insert(condition.get_pair().var);
+            }
+        }
+
+        for (EffectProxy effect: op.get_effects()) {
+            for (FactProxy condition: effect.get_conditions()) {
+                if (condition.get_variable().is_derived() && condition.get_value() == 1) {  // TODO: fix the hardcoded 1
+                    needed_negatively.insert(condition.get_pair().var);
+                }
+            }
+        }
+    }
+    unordered_map<int, vector<int>> positive_dependencies;
+    unordered_map<int, vector<OperatorProxy>> axioms_for_var;
+    for (OperatorProxy axiom: task_proxy.get_axioms()) {
+        int head_var = axiom.get_effects()[0].get_fact().get_variable().get_id();
+        if (axioms_for_var.count(head_var) > 0) {
+            axioms_for_var[head_var].push_back(axiom);
+        } else {
+            axioms_for_var.insert({head_var, {axiom}});
+        }
+        for (FactProxy condition: axiom.get_effects()[0].get_conditions()) {
+            if (condition.get_variable().is_derived()) {
+                int var = condition.get_variable().get_id();
+                if (condition.get_value() == 1) { // TODO: fix the hardcoded 1
+                    needed_negatively.insert(var);
+                } else {
+                    if (positive_dependencies.count(head_var) > 0) {
+                        positive_dependencies[head_var].push_back(var);
+                    } else {
+                        positive_dependencies.insert({head_var, {var}});
+                    }
+                }
+            }
+        }
+    }
+
+    // All derived variables that occur positively in an axiom for a needed negatively var are also needed negatively.
+    deque<int> to_process(needed_negatively.begin(), needed_negatively.end());
+    while (!to_process.empty()) {
+        int var = to_process.front();
+        to_process.pop_front();
+        for (int var2: positive_dependencies[var]) {
+            if (needed_negatively.count(var2) == 0) {
+                needed_negatively.insert(var2);
+                to_process.push_back(var2);
+            }
+        }
+    }
+
+    // If there are no derived variables that occur as their default value we don't need to do anything.
+    if (needed_negatively.empty()) {
+        return;
+    }
+
+    // Build dependency graph and collect relevant axioms for derived values occurring as their default value.
+    // TODO: The dependency information is a superset of positive_dependencies from above.
```

Comment:
I resolved this by only using positive_dependencies, since negative dependencies cannot introduce a cycle (this would imply non-stratifiable axioms, which is already checked in the translator.)